### PR TITLE
Adding Recommendation data structure and RecommendationRepo

### PR DESCRIPTION
**Pull request by lawzlo**
@stkem @yingjieMiao @leogrim

### CODE REVIEW

**Comment by lawzlo** [comment]
I separate recommendation to AdminRecommendation(for testing) and Recommendation(for all users). We can remove AdminRecommendation later after finishing  pre-generating recommendation feature.

File: kifi-backend/modules/abook/test/com/keepit/abook/model/FriendAdminRecommendationTest.scala [hunk 1]
```diff
@@ -4,7 +4,7 @@ import org.specs2.mutable.Specification
 import com.keepit.abook.ABookTestInjector
 import com.keepit.common.db.Id
 
-class FriendRecommendationTest extends Specification with ABookTestInjector {
+class FriendAdminRecommendationTest extends Specification with ABookTestInjector {
```
Comment (stkem) [inline kifi-backend/modules/abook/test/com/keepit/abook/model/FriendAdminRecommendationTest.scala]: I think this is unrelated to the Curator recommendations.

File: kifi-backend/modules/common/app/com/keepit/curator/model/AdminRecommendation.scala [hunk 2]
```diff
@@ -5,7 +5,7 @@ import com.keepit.model.{ NormalizedURI, User }
 
 import com.kifi.macros.json
 
-@json case class Recommendation(
+@json case class AdminRecommendation(
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/curator/model/AdminRecommendation.scala]: This is effectively a serialization format for the `Recommendation` database model, it won't necessarily be just for admins. A naming pattern we use for this is `<DB Model name> Info`, so in this case `RecommendationInfo`.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.keepit.curator.model
+
+import com.keepit.common.db._
+import com.keepit.common.time._
+import com.keepit.model.{ User, NormalizedURI }
+import org.joda.time.DateTime
+import play.api.libs.json.JsObject
+
+case class Recommendation(
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Something I forgot to mention on Friday: This should also have an `externalId` (ultimately to be used for de-duplication on the client when needed). Look at the `ModelWithExternalId` trait and it's users for examples on how to add this.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    createdAt: DateTime = currentDateTime,
+    updateAt: DateTime = currentDateTime,
+    uriId: Id[NormalizedURI],
+    userId: Id[User],
+    finalScore: Float,
+    allScores: JsObject,
+    seen: Boolean,
+    clicked: Boolean,
+    kept: Boolean,
+    state: State[Recommendation] = RecommendationStates.ACTIVE) extends Model[Recommendation] {
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Not a big thing, but I find it helps readability to group the fields with default values together at the beginning with the class definition. 
Also, if you inherit from `ModelWithState` the compiler will make sure the state is correct.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import play.api.libs.json.JsObject
+
+case class Recommendation(
+    id: Option[Id[Recommendation]] = None,
+    createdAt: DateTime = currentDateTime,
+    updateAt: DateTime = currentDateTime,
+    uriId: Id[NormalizedURI],
+    userId: Id[User],
+    finalScore: Float,
+    allScores: JsObject,
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: This should be of type `UriScores` not `JsObject`, just the storage should be Json. You can use slick Type Mappers to make that work. I can walk you though it when you get to it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import org.joda.time.DateTime
+import play.api.libs.json.JsObject
+
+case class Recommendation(
+    id: Option[Id[Recommendation]] = None,
+    createdAt: DateTime = currentDateTime,
+    updateAt: DateTime = currentDateTime,
+    uriId: Id[NormalizedURI],
+    userId: Id[User],
+    finalScore: Float,
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: How about `masterScore` or something like that. Nothing is every final :-)

File: kifi-backend/modules/common/app/com/keepit/curator/model/AdminRecommendation.scala (hunk 2 above)
Comment (lawzlo) [inline kifi-backend/modules/common/app/com/keepit/curator/model/AdminRecommendation.scala]: Renaming it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 3 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Cool. Will add it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 6 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Sure. `masterScore` or `recommendScore`, which one do you think is better?

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 5 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Cool. Will change it.

File: kifi-backend/modules/abook/test/com/keepit/abook/model/FriendAdminRecommendationTest.scala (hunk 1 above)
Comment (lawzlo) [inline kifi-backend/modules/abook/test/com/keepit/abook/model/FriendAdminRecommendationTest.scala]: I use IDE to refactor all names, will check it.

Comment (lawzlo) [inline kifi-backend/modules/abook/test/com/keepit/abook/model/FriendAdminRecommendationTest.scala]: seems it's unrelated. Renaming it back.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+@ImplementedBy(classOf[RecommendationRepoImpl])
+trait RecommendationRepo extends DbRepo[Recommendation] {
+  def getByUriId(uriId: Id[NormalizedURI])(implicit session: RSession): Seq[Recommendation]
+  def getByUserId(userId: Id[User])(implicit session: RSession): Seq[Recommendation]
+  def getByUriAndUserId(uriId: Id[NormalizedURI], userId: Id[User])(implicit session: RSession): Option[Recommendation]
+  def getFirstByUriId(uriId: Id[NormalizedURI])(implicit session: RSession): Option[Recommendation]
+  def getByTopFinalScore(userId: Id[User], maxBatchSize: Int)(implicit session: RSession): Seq[Recommendation]
+  def deactivateRecommendationFromUser(uriId: Id[NormalizedURI], userId: Id[User])(implicit session: RSession): Boolean
+}
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala]: I don't think we need any of these methods except `getByUriAndUserId` and `getByTopFinalScore`. As a rule we try to add repo methods only when we really need them and in conjuction with the SQL evolution to avoid accidentally introducing methods that look simple but are actually really expensive because of the way the db schema is set up.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  type RepoImpl = RecommendationTable
+
+  class RecommendationTable(tag: Tag) extends RepoTable[Recommendation](db, tag, "recommendation_item") {
+    def uriId = column[Id[NormalizedURI]]("uri_id", O.NotNull)
+    def userId = column[Id[User]]("user_id", O.NotNull)
+    def finalScore = column[Float]("final_score", O.NotNull)
+    def allScore = column[JsObject]("all_score", O.NotNull)
+    def seen = column[Boolean]("was_seen", O.NotNull)
+    def clicked = column[Boolean]("was_clicked", O.NotNull)
+    def kept = column[Boolean]("was_kept", O.NotNull)
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala]: `seen`, `clicked` and `kept` are fine here, I think. To keep the column names the same as the actual field names (general patterns we use is to simple convert from CamelCase to snake_case)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 6 above)
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: I like `masterScore`

**Comment by stkem** [comment]
Good work. Made some comments. Also, you are missing the evolution (.sql) file.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala (hunk 8 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala]: Cool. Renaming it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 6 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Deal.

**Comment by lawzlo** [comment]
Good call for evolution file. Will add it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala (hunk 7 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/RecommendationRepo.scala]: Hmm...Maybe we don't need getByUriId, getByUserId or getFirstByUriId for now. Do we need deactivateRecommendationFromUser? This will help us to manually hide some recommendation from user if RecommendationRestriction doesn't filter it out successfully?

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 4 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Cool. Fixing it.

Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: :-)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: We may want a PublicId instead? @andrewconner are we trying to move away from external ids altogether?

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 4 above)
Comment (leogrim) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: We may want to call this ItemRecommendation or UriRecommendation, so that there's no conflict with other types of recommendations (I'm thinking FriendRecommendation from PYMK)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 3 above)
Comment (andrewconner) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: When it makes sense, yep. We need to document best practices, but in general, when you need an ID that:
- Is used by clients only, never intra-service
- Is often converted from Id -> external ID, or vice versa.
- Does not give user authorization to access trusted resource (because `PublicId`s are not cryptographically secure due to the need for unchanging key)

go for it

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 4 above)
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Good point. My vote goes to `UriRecommendation`.

Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Ya, `UriRecommendation` is more specific and good name for curator service layer. Will rename it.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala (hunk 3 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/Recommendation.scala]: Will change to use PublicId. Thanks.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    def clicked = column[Boolean]("clicked", O.NotNull)
+    def kept = column[Boolean]("kept", O.NotNull)
+    def * = (id.?, createdAt, updatedAt, state, uriId, userId, masterScore, allScore, seen, clicked, kept) <> ((UriRecommendation.apply _).tupled, UriRecommendation.unapply _)
+  }
+
+  def table(tag: Tag) = new RecommendationTable(tag)
+  initTable()
+
+  def getByUriAndUserId(uriId: Id[NormalizedURI], userId: Id[User])(implicit session: RSession): Option[UriRecommendation] = {
+    (for (row <- rows if row.uriId === uriId && row.userId === userId && row.state === UriRecommendationStates.ACTIVE) yield row).firstOption
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: I think this should not filter by state, since we may want to get an inactive item and update it to be active if there is new information, for example.

Comment (andrewconner) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: A common pattern is an option param for states to filter out.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/SeedItem.scala [hunk 10]
```diff
@@ -22,6 +21,34 @@ case class SeedItem(
   lastSeen: DateTime,
   keepers: Keepers)
 
+object UriScoresSerializer extends Format[UriScores] {
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/SeedItem.scala]: I should have mentioned this earlier, but we have a nice macro (which martin wrote) that will generate these for you (and keep them updated). You can just add `@json` in fornt of the `UriScores` class definition and it will magically work :-)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+class UriRecommendationRepoImpl @Inject() (
+  val db: DataBaseComponent,
+  val clock: Clock)
+    extends DbRepo[UriRecommendation] with UriRecommendationRepo with Logging {
+
+  import DBSession._
+  import db.Driver.simple._
+
+  implicit val uriScoresMapper = MappedColumnType.base[UriScores, String](UriScoresSerializer.writes(_).toString, s => UriScoresSerializer.reads(Json.parse(s)).get)
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: If you use the macro you'll have to use `Json.toJson` instead of `UriScoresSerializer.writes` and `Json.fromJson` instead of `UriScoresSerializer.reads`. 
Also, the `toString` method is generally intended to produce something for human consumption. In this case that works correctly but it might change at any time. Better to use `Json.stringify`

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala [hunk 12]
```diff
@@ -21,7 +21,7 @@ class RecommendationGenerationCommander @Inject() (
 
   val defaultScore = 0.0f
 
-  def getAdHocRecommendations(userId: Id[User], howManyMax: Int, scoreCoefficients: Map[ScoreType.Value, Float]): Future[Seq[Recommendation]] = {
+  def getAdHocAdminRecommendations(userId: Id[User], howManyMax: Int, scoreCoefficients: Map[ScoreType.Value, Float]): Future[Seq[RecommendationInfo]] = {
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala]: No need for the word `admin` in here any more, right? (happens to be admin only right now, but doesn't need to be)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  def table(tag: Tag) = new RecommendationTable(tag)
+  initTable()
+
+  def getByUriAndUserId(uriId: Id[NormalizedURI], userId: Id[User])(implicit session: RSession): Option[UriRecommendation] = {
+    (for (row <- rows if row.uriId === uriId && row.userId === userId && row.state === UriRecommendationStates.ACTIVE) yield row).firstOption
+  }
+
+  def getByTopFinalScore(userId: Id[User], maxBatchSize: Int)(implicit session: RSession): Seq[UriRecommendation] = {
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: `getByTopMasterScore`

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+  implicit val uriScoresMapper = MappedColumnType.base[UriScores, String](UriScoresSerializer.writes(_).toString, s => UriScoresSerializer.reads(Json.parse(s)).get)
+
+  type RepoImpl = RecommendationTable
+
+  class RecommendationTable(tag: Tag) extends RepoTable[UriRecommendation](db, tag, "uri_recommendation_item") {
+    def uriId = column[Id[NormalizedURI]]("uri_id", O.NotNull)
+    def userId = column[Id[User]]("user_id", O.NotNull)
+    def masterScore = column[Float]("master_score", O.NotNull)
+    def allScore = column[UriScores]("all_score", O.NotNull)
```
Comment (stkem) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: `allScores` (with an s) is probably clearer

File: kifi-backend/modules/common/conf/evolutions/shoebox/213.sql [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    uri_id bigint(20) NOT NULL,
+    master_score(10) float NOT NULL,
+    all_score varchar(256) NOT NULL,
+    seen boolean NOT NULL,
+    clicked boolean NOT NULL,
+    kept boolean NOT NULL,
+
+    PRIMARY KEY (id),
+
+    UNIQUE INDEX recommendation_u_user_id_uri_id (user_id, uri_id)
```
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/213.sql]: We are also going to need an index on `(user_id, master_score)` to make the top item look up fast.

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala (hunk 9 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: Will fix.

File: kifi-backend/modules/common/conf/evolutions/shoebox/213.sql (hunk 15 above)
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/213.sql]: In which case I would reverse the other index (i.e. -> `(uri_id, user_id)`) to get fast lookups by `uri_id` for free, which we will need soon for renormalizations (i.e. when a uri_id changes)

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala (hunk 14 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: fixing

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala (hunk 13 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: fixing

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala (hunk 11 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: fixing

File: kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala (hunk 12 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/commanders/RecommendationGenerationCommander.scala]: ya, renaming back.

File: kifi-backend/modules/common/conf/evolutions/shoebox/213.sql (hunk 15 above)
Comment (lawzlo) [inline kifi-backend/modules/common/conf/evolutions/shoebox/213.sql]: Fixing

File: kifi-backend/modules/curator/app/com/keepit/curator/model/SeedItem.scala (hunk 10 above)
Comment (lawzlo) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/SeedItem.scala]: using martin's json annotation.

**Comment by stkem** [comment]
:+1:
